src/dwarf.rs

```diff
@@ -1483,69 +1480,27 @@ mod tests {
-        assert!(resolver_r.is_ok());
-        let resolver = resolver_r.unwrap();
+        let bin_name = Path::new(&env!("CARGO_MANIFEST_DIR"))
+            .join("data")
+            .join("test.bin");
+        let resolver = DwarfResolver::open(bin_name.as_ref(), true, false).unwrap();
         let (addr, dir, file, line) = resolver.pick_address_for_test();
 
-        let line_info = resolver.find_line(addr);
-        assert!(line_info.is_some());
-        let (dir_ret, file_ret, line_ret) = line_info.unwrap();
-        println!("{}/{} {}", dir_ret, file_ret, line_ret);
+        let (dir_ret, file_ret, line_ret) = resolver.find_line(addr).unwrap();
         assert_eq!(dir, dir_ret);
         assert_eq!(file, file_ret);
         assert_eq!(line, line_ret);
     }
 
     #[test]
     fn test_debug_info_parse_symbols() {
-        let bin_name = env::args().next().unwrap();
-        let parser_r = Elf64Parser::open(bin_name.as_ref());
-        assert!(parser_r.is_ok());
-        let parser = parser_r.unwrap();
+        let bin_name = Path::new(&env!("CARGO_MANIFEST_DIR"))
+            .join("data")
+            .join("test.bin");
 
-        let result = debug_info_parse_symbols(&parser, None, 4);
-
-        assert!(result.is_ok());
-        let syms = result.unwrap();
-
-        let mut myself_found = false;
-        let mut myself_addr: u64 = 0;
-        let mut parse_symbols_found = false;
-        let mut parse_symbols_addr: u64 = 0;
-        for sym in syms {
-            if sym
-                .name
-                .starts_with("_ZN8blazesym5dwarf5tests29test_debug_info_parse_symbols17h")
-            {
-                myself_found = true;
-                myself_addr = sym.address;
-            } else if sym
-                .name
-                .starts_with("_ZN8blazesym5dwarf24debug_info_parse_symbols17h")
-            {
-                parse_symbols_found = true;
-                parse_symbols_addr = sym.address;
-            }
-        }
-        assert!(myself_found);
-        assert!(parse_symbols_found);
-        assert_eq!(
-            (test_debug_info_parse_symbols as fn() as *const fn() as i64)
-                - (debug_info_parse_symbols
-                    as for<'a> fn(
-                        &'a Elf64Parser,
-                        Option<&(dyn Fn(&DWSymInfo<'a>) -> bool + Send + Sync)>,
-                        usize,
-                    ) -> Result<Vec<DWSymInfo<'a>>, Error>
-                    as *const (
-                        &'_ Elf64Parser,
-                        Option<&(dyn Fn(&DWSymInfo<'_>) -> bool + Send + Sync)>,
-                        usize,
-                    ) as i64),
-            myself_addr as i64 - parse_symbols_addr as i64
-        );
+        let parser = Elf64Parser::open(bin_name.as_ref()).unwrap();
+        let syms = debug_info_parse_symbols(&parser, None, 4).unwrap();
+        assert!(syms.iter().any(|sym| sym.name == "fibonacci"))
```

Comment:
This is a simpler sample compared to the binary of the test case harness.  I suggest having two separate tests.  The old one against the binary of the harness explores potential errors in a complicated dynamic environment.  The new one ensures the basic functionalities.

Reply:
You'll have to define this "complicated dynamic environment" is and then we can work out a way to test that. The test as it was written tested the existence of two symbols from what I can tell, and we test for one. I am happy to extend it to two if that's what you are asking. Whether there was more stuff in the binary or not that could potentially trigger other code paths was not mentioned anywhere, nor checked for, nor truly guaranteed. Let's stop making random assumptions. If that's what should be tested then that has to be spelled out and defined, at the very least in the test's name. In any case, I don't see a regression here given the above.

Reply:
Got it!